src/levanter/main/sft_mixture.py

```diff
@@ -0,0 +1,306 @@
+import asyncio
+import dataclasses
+import logging
+import os
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Dict, Iterator, Optional, Union
+
+import jax.random as jrandom
+import transformers
+
+import haliax as hax
+from haliax import Axis
+from haliax.partitioning import round_axis_for_partitioning
+
+import levanter
+from levanter import callbacks
+from levanter.compat.hf_checkpoints import HFCheckpointConverter, HFCompatConfig, save_hf_checkpoint_callback
+from levanter.data import PermutationDataset
+from levanter.data.dataset import AsyncDataset
+from levanter.data.loader import stack_batches
+from levanter.data.mixture import MixtureDataset, StopStrategy
+from levanter.data.packing import PromptCompletion, pack_prompt_completions
+from levanter.data.text import (
+    ChatUrlDataSourceConfig,
+    SupervisedSourceConfig,
+    mk_cached_sft_dataset,
+    mk_supervised_dataset,
+)
+from levanter.models.llama import LlamaConfig
+from levanter.models.lm_model import LmConfig, LmHeadModel, compute_next_token_loss
+from levanter.optim import AdamConfig, OptimizerConfig
+from levanter.trainer import Trainer, TrainerConfig
+from levanter.utils.background_iterable import BackgroundIterator
+
+
+logger = logging.getLogger(__name__)
+
+# Define default special tokens
+DEFAULT_PAD_TOKEN = "[PAD]"
+DEFAULT_EOS_TOKEN = "</s>"
+DEFAULT_BOS_TOKEN = "<s>"
+DEFAULT_UNK_TOKEN = "<unk>"
+
+
+class DatasetType(str, Enum):
+    """Type of dataset to use"""
+
+    HUGGINGFACE = "huggingface"  # Use HF dataset
+    CHAT_JSONL = "chat_jsonl"  # Use JSONL files with chat format
+
+
+@dataclass
+class SFTMixtureConfig:
```

Comment:
i'd prefer if we had one main instead of this, but let's save that for the pre-train/sft merger